server/__tests__/calculation.test.js

```diff
@@ -0,0 +1,122 @@
+} = require("../_jest-setup_/utils/server-setup");
+
+let server;
+
+beforeAll(async () => {
+  server = await setupServer();
+});
+
+afterAll(async () => {
+  await teardownServer();
+});
+
+describe("Calculation API Endpoints", () => {
+  //////////////////////////////
+  //          general          //
+  //////////////////////////////
+  let calcId; // id of the calculation
+
+  // GET "/" all calculations
+  it("should get all calculations", async () => {
+    const res = await request(server).get("/api/calculations");
+    expect(Array.isArray(res.body)).toBeTruthy();
+    res.body.forEach(calc => {
+      expect(calc).toHaveProperty("id");
+      expect(calc).toHaveProperty("name");
+      expect(calc).toHaveProperty("description");
+      expect(calc).toHaveProperty("deprecated");
+      expect(calc).toHaveProperty("dateCreated");
+      expect(calc).toHaveProperty("dateModified");
+    });
+    calcId = res.body[0].id;
+  });
+
+  // GET "/:calcId" calculation by id
+  it("should get calculation by id", async () => {
+    const res = await request(server).get(`/api/calculations/${calcId}`);
+    expect(res.statusCode).toEqual(200);
+    expect(res.body.id).toEqual(calcId);
+  });
+
+  // GET "/:calcId" calculation by inexistent id
+  it("should get calculation by id", async () => {
+    const res = await request(server).get(`/api/calculations/9999999`);
+    expect(res.statusCode).toEqual(404);
+  });
+
+  // GET "/:calcId/rules" all rules for a calculation
+  it("should get all rules for a calculation", async () => {
+    const res = await request(server).get(`/api/calculations/${calcId}/rules`);
+    expect(res.statusCode).toEqual(200);
+    expect(Array.isArray(res.body)).toBeTruthy();
+    res.body.forEach(calc => {
+      expect(calc).toHaveProperty("id");
+      expect(calc).toHaveProperty("calculationId");
+      expect(calc.calculationId).toEqual(calcId);
+    });
+  });
+
+  //////////////////////////////
+  //      admin endpoints      //
+  //////////////////////////////
+  let adminToken; // token of the admin
+  let newCalcId; // id of the new calculation
+
+  // POST "accounts/login/:email?" Login as admin to get token
+  it("should login as a security admin", async () => {
+    const res = await request(server).post("/api/accounts/login").send({
+      email: process.env.ADMIN_EMAIL,
+      password: process.env.ADMIN_PASSWORD
+    });
+    expect(res.statusCode).toEqual(200);
+    expect(res.body).toHaveProperty("token");
+    adminToken = res.body.token;
+  });
+
+  // POST "/" Create a calculation (Admin only)
+  it("should create a calculation", async () => {
+    const res = await request(server)
+      .post("/api/calculations")
+      .set("Authorization", `Bearer ${adminToken}`)
+      .send({
+        name: "Test Name",
+        description: "Test Description",
+        deprecated: false
+      });
+    expect(res.statusCode).toEqual(201);
+    expect(res.body).toHaveProperty("id");
+    newCalcId = res.body.id;
```

Comment:
expect()?

Reply:
line 92 intentional - this sets the variable that is used in the following tests

Reply:
Other tests depend on this side effect? Tests need to runnable as a single unit.

Reply:
It's intentional for the nature of these tests since it avoids mocking - everything tries to be as "real" as possible. We discussed adding functional tests that would be as close to the production environment as possible. So certain tests have protected endpoints which require JWT's

We had discussed trying to add unit tests but this way added more coverage and better ROI on time for where we are in the project.

Thoughts?

Reply:
For example, jest can rerun just failed tests or based on regex. As your tests scale, running subsets and reordering will become even more important. 

Don't let tests depend on other tests.

Reply:
> If the tests must run in a certain order they should just be a single test.
I put it in a single 'describe' block with this intention - is that pattern inappropriate? 

> Don't let tests depend on other tests.
Could this be done without mocking?

Reply:
Describe is any arbitrary semantic group. "It" is the test. 

So you could group tests by all needing the same before/after assumptions. 

```
describe ('tests that need an calculation')
  let calcId = null
  beforeEach() { calcId = ... }
  afterEach() { cleanup(calcId)}
  it('verifies delete') { 
    delete(calcId);
    expect(get(calcId)).toBeFalsy()
  }
  it('verifiers update') ...
```

So each test starts with clean assumptions.

Reply:
Just refactored based on this suggestion

commit 7f1deff8af2698b17763ceb17ab8f937677fc188